components/brave_shields/content/browser/ad_block_service.cc

```diff
@@ -209,6 +208,41 @@ base::Value::Dict AdBlockService::UrlCosmeticResources(
         }
       }
     }
+
+    base::Value::List* procedural_actions =
+        resources.FindList("procedural_actions");
+    // In standard blocking mode, drop procedural filters but otherwise keep
+    // action filters.
+    if (procedural_actions) {
+      base::Value::List::iterator it = procedural_actions->begin();
+      while (it < procedural_actions->end()) {
+        DCHECK(it->is_string());
+        auto* pfilter_str = it->GetIfString();
+        if (pfilter_str == nullptr) {
+          continue;
+        }
+        auto val = base::JSONReader::ReadDict(*pfilter_str);
```

Comment:
this is json embedded as a string? Is there a link or doc somewhere describing the format of this?

Reply:
We have a bunch of unit tests for it in `adblock-rust`, and the format is documented at https://github.com/brave/adblock-rust/blob/next/src/cosmetic_filter_cache.rs#L384-L399 (will be on docs.rs once the version bump is officially published)

Reply:
but this parsing is happening in AdBlockService and I don't see any unit tests here for this part

Reply:
and can you link to the docs for the formatting here in comments?

Reply:
split it into `adblock_service_helper`; added the link to the docs from there and some unit tests